src/mlpack/methods/ann/layer/multihead_attention_impl.hpp

```diff
@@ -0,0 +1,288 @@
+/**
+ * @file methods/ann/layer/multihead_attention_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of the MultiheadAttention class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_MULTIHEAD_ATTENTION_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_MULTIHEAD_ATTENTION_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "multihead_attention.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template <typename InputDataType, typename OutputDataType,
+          typename RegularizerType>
+MultiheadAttention<InputDataType, OutputDataType, RegularizerType>::
+MultiheadAttention() :
+    tLen(0),
+    sLen(0),
+    embedDim(0),
+    numHeads(0),
+    headDim(0),
+    dropoutRate(0.0),
+    deterministic(false)
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType,
+          typename RegularizerType>
+MultiheadAttention<InputDataType, OutputDataType, RegularizerType>::
+MultiheadAttention(const size_t tLen,
+    const size_t sLen,
+    const size_t embedDim,
+    const size_t numHeads,
+    const ElemType dropoutRate,
+    const bool deterministic) :
+    tLen(tLen),
+    sLen(sLen),
+    dropoutRate(dropoutRate),
+    deterministic(deterministic)
+{
+  Log::Assert(embedDim % numHeads == 0,
+      "'embedDim' must be divisible by 'numHeads'");
+  this->embedDim = embedDim;
+  this->numHeads = numHeads;
+  this->headDim = embedDim / numHeads;
+
+  weights.set_size(4 * (embedDim + 1) * embedDim, 1);
+  dropout.Ratio(dropoutRate);
+  dropout.Deterministic() = deterministic;
+}
+
+template <typename InputDataType, typename OutputDataType,
+          typename RegularizerType>
+void MultiheadAttention<InputDataType, OutputDataType, RegularizerType>::
+ResetParameters()
+{
+  queryWt = arma::mat(weights.memptr(), embedDim, embedDim, false, false);
+  keyWt = arma::mat(weights.memptr() + embedDim * embedDim,
+      embedDim, embedDim, false, false);
+  valueWt = arma::mat(weights.memptr() + 2 * embedDim * embedDim,
+      embedDim, embedDim, false, false);
+  outWt = arma::mat(weights.memptr() + 3 * embedDim * embedDim,
+      embedDim, embedDim, false, false);
+
+  queryBias = arma::mat(weights.memptr()
+      + 4 * embedDim * embedDim, embedDim, 1, false, false).zeros();
+  keyBias = arma::mat(weights.memptr()
+      + (4 * embedDim + 1) * embedDim, embedDim, 1, false, false).zeros();
+  valueBias = arma::mat(weights.memptr()
+      + (4 * embedDim + 2) * embedDim, embedDim, 1, false, false).zeros();
+  outBias = arma::mat(weights.memptr()
+      + (4 * embedDim + 3) * embedDim, embedDim, 1, false, false).zeros();
```

Comment:
Does it work with `float`s? I guess in this case `weights.memptr()` should return `float*`, but `arma::mat()` requires `double*`. Perhaps there should be `arma::Mat<TheSameElemType>()`.

Reply:
Yes, you are right. I will handle the templating once the implementation looks good. It was just that my ide was getting cluttered. :smile:

Reply:
Done.